src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+            return sessionToken;
+        }
+
+        private static async Task<string> GetEC2Response(Uri baseUri)
+        {
+            // Acquire token
+
+            var tokenRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, "latest/api/token"),
+                Method = HttpMethod.Put,
+            };
+            tokenRequest.Headers.Add("X-aws-ec2-metadata-token-ttl-seconds", "30");
+            var token = await GetHttpContent(tokenRequest, "Failed to acquire EC2 token.").ConfigureAwait(false);
+
+            // Acquire role name
+
+            var roleRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/"),
+                Method = HttpMethod.Get
+            };
+            roleRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var roleName = await GetHttpContent(roleRequest, "Failed to acquire EC2 role name.").ConfigureAwait(false);
+
+            // Acquire credentials
+
+            var credentialsUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/");
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(credentialsUri, roleName),
+                Method = HttpMethod.Get
+            };
+            credentialsRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var credentials = await GetHttpContent(credentialsRequest, "Failed to acquire EC2 credentials.").ConfigureAwait(false);
+
+            return credentials;
+        }
+
+        private static async Task<string> GetECSResponse(Uri baseUri, string relativeUri)
+        {
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, relativeUri),
+                Method = HttpMethod.Get
+            };
+
+            return await GetHttpContent(credentialsRequest, "Failed to acquire ECS credentials.").ConfigureAwait(false);
+        }
+
+        private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
+        {
+            HttpResponseMessage response;
+            try
+            {
+                response = await _httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            catch (HttpRequestException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new MongoClientException(exceptionMessage);
+            }
+
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        private static void ValidateCredentials(string username, string password, string sessionToken)
```

Comment:
this can be removed via the suggestion here: https://github.com/DmitryLukyanov/mongo-csharp-driver/commit/1b03be95b0bd47b12600bc634bcbe112da1d866f#diff-e16c900aa981991f12bb138991116445R34